tools/Customization/DevHome.Customization/Views/AddRepositoriesView.xaml.cs

```diff
@@ -37,16 +38,46 @@ public void RemoveFolderButton_Click(object sender, RoutedEventArgs e)
         }
     }
 
-    public void AssignSourceControlProviderButton_Click(object sender, RoutedEventArgs e)
+    public async void AssignSourceControlProviderButton_Click(object sender, RoutedEventArgs e)
     {
         // Extract relevant data from view and give to view model for assign
         MenuFlyoutItem menuItem = (MenuFlyoutItem)sender;
         if (menuItem.DataContext is RepositoryInformation repoInfo)
         {
-            ViewModel.AssignSourceControlProviderToRepository(menuItem.Text, repoInfo.RepositoryRootPath);
+            var taskResult = await ViewModel.AssignSourceControlProviderToRepository(menuItem.Text, repoInfo.RepositoryRootPath);
+            if (taskResult?.Result != Helpers.ResultType.Success)
+            {
+                _log.Error("Error occurred while assigning source control provider: ", taskResult?.Error, taskResult?.Exception, taskResult?.DiagnosticText, taskResult?.DisplayMessage);
+                ShowErrorContentDialog(taskResult!, this.XamlRoot);
+            }
         }
     }
 
+    public async void ShowErrorContentDialog(SourceControlValidationResult result, XamlRoot xamlRoot)
+    {
+        var stringResource = new StringResource("DevHome.Customization.pri", "DevHome.Customization/Resources");
+        string diagnosticText = result.DiagnosticText ?? string.Empty;
+        string content;
+        if (diagnosticText.Contains("Invalid repository path"))
+        {
+            content = stringResource.GetLocalized("AssignSourceControlErrorDialogInvalidRepositoryPathContent");
+        }
+        else
+        {
+            content = string.Concat(stringResource.GetLocalized("AssignSourceControlErrorDialogDefaultContent"), diagnosticText);
```

Comment:
Why display `DiagnosticText`? Isn't that what `DisplayMessage` is for?

Reply:
The DisplayMessage being provided by the extension is a catch all message and the diagnostic text seemed to provide much more valuable information that may be of interest to a developer audience.  If we would still like to show the DisplayMessage, we can alter the extension to catch some common LibGit2Sharp exceptions (such as one that occurs when a repository path does not point to a valid git repository) and send more distinct display messages in the GetLocalRepositoryResult. Any thoughts?

Reply:
This should be documented better, (maybe even wrapped up and standardized in the SDK into a helper class that holds an Exception, DisplayMessage, and DiagnosticText).

The convention that seems to domincate is:
- `DisplayMessage` is for localized content to display to the user.
- `DiagnosticText` is for non-localized content that will be sent back solely for Dev Home to process (most likely in logs and/or telemetry).

> catch some common LibGit2Sharp exceptions (such as one that occurs when a repository path does not point to a valid git repository) and send more distinct display messages in the GetLocalRepositoryResult.

That sounds perfect! The LibGit2Sharp exceptions are less likely to be useful in the context of our extension communicating back. We should catch any exceptions coming out of the Git extension and translate them into something meaningful to DevHome (ideally actionable). This also makes sense if we continue to migrate functionality away from LibGit2Sharp, as we'll want to originate our own, more specific error codes and messages.

I'm pretty sure libgit2 (the underlying library used by libgit2sharp) does not localize its error messages. So we won't be able to reuse the error messages in the exceptions coming from libgit2sharp. What's worse, is that libgit2sharp doesn't seem to preserve the libgit2 error code when it transforms the error into an exception, so trying to catch anything that doesn't have a dedicated exception type will be fraught with the perils of string matching and trying to translate into a better exception to send to DevHome. So far off the top of my head, some of the exceptions we are likely to see in our scenarios are:
- `LibGit2Sharp.RepositoryNotFoundException` seems self-explanatory. This is can result directly from `Repository.Open`.
- The file ownership error (e.g. trying to open a repo directly in a WSL path without invoking wsl). libgit2 sets the error to `GIT_EOWNER` (-36), but libgit2sharp turns this into a generic `LibGit2Sharp.LibGit2SharpException` with the message set to whatever libgit2 said. Matching it is brittle and prone to break in the future, but at this *particular moment* it is `repository path '%.*s' is not owned by current user` [source](https://github.com/libgit2/libgit2/blob/782e29c906f6e44b120843356f286b6a97d89f88/src/libgit2/repository.c#L713)
  - Alternatively, if we receive this error from git.exe, it seems to print the following. Alas, git.exe also does not return error codes. I've found other git-related code in the wild that specifically searches for part of this string to detect this case.
```
fatal: detected dubious ownership in repository at '%s'.
To add an exception for this directory, call:

        git config --global --add safe.directory '%s'
```

Reply:
Pushed new changes for the above. Thanks!